flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/sql/join/JoinReorderITCase.java

```diff
@@ -0,0 +1,465 @@
+        List<Row> results =
+                CollectionUtil.iteratorToList(
+                        DataStreamUtils.collect(
+                                tEnv.toAppendStream(tEnv.sqlQuery(query), Row.class)));
+        String expected = "+I[Hallo, Hi, Hallo, Hallo]";
+        TestBaseUtils.compareResultAsText(results, expected);
+    }
+
+    @Test
+    public void testBusyTreeJoinReorder() throws Exception {
+        CatalogColumnStatisticsDataLong longColStats =
+                new CatalogColumnStatisticsDataLong(100L, 100L, 50L, 1000L);
+        Map<String, CatalogColumnStatisticsDataBase> colStatsMap = new HashMap<>(1);
+        colStatsMap.put("b1", longColStats);
+        catalog.alterTableColumnStatistics(
+                new ObjectPath(tEnv.getCurrentDatabase(), "T1"),
+                new CatalogColumnStatistics(colStatsMap),
+                false);
+
+        longColStats = new CatalogColumnStatisticsDataLong(100L, 100L, 500000L, 1000L);
+        colStatsMap = new HashMap<>(1);
+        colStatsMap.put("b2", longColStats);
+        catalog.alterTableColumnStatistics(
+                new ObjectPath(tEnv.getCurrentDatabase(), "T2"),
+                new CatalogColumnStatistics(colStatsMap),
+                false);
+
+        longColStats = new CatalogColumnStatisticsDataLong(100L, 100L, 50L, 1000L);
+        colStatsMap = new HashMap<>(1);
+        colStatsMap.put("b3", longColStats);
+        catalog.alterTableColumnStatistics(
+                new ObjectPath(tEnv.getCurrentDatabase(), "T3"),
+                new CatalogColumnStatistics(colStatsMap),
+                false);
+
+        longColStats = new CatalogColumnStatisticsDataLong(100L, 100L, 500000L, 1000L);
+        colStatsMap = new HashMap<>(1);
+        colStatsMap.put("b4", longColStats);
+        catalog.alterTableColumnStatistics(
+                new ObjectPath(tEnv.getCurrentDatabase(), "T4"),
+                new CatalogColumnStatistics(colStatsMap),
+                false);
+
+        String query =
+                "SELECT tab2.d4, tab2.c3, tab1.d2, tab1.d1 FROM "
+                        + "(SELECT * FROM T1 JOIN T2 ON T1.b1 = T2.b2) tab1 "
+                        + "JOIN (SELECT * FROM T3 JOIN T4 ON T3.b3 = T4.b4) tab2 "
+                        + "ON tab1.b2 = tab2.b4";
+
+        List<Row> results =
+                CollectionUtil.iteratorToList(
+                        DataStreamUtils.collect(
+                                tEnv.toAppendStream(tEnv.sqlQuery(query), Row.class)));
+        String expected =
+                "+I[Hallo Welt, Hello, Hallo Welt, Hallo Welt]\n"
+                        + "+I[Hallo Welt, Hello world, Hallo Welt, Hallo Welt]\n"
+                        + "+I[Hallo, Hi, Hallo, Hallo]";
+        TestBaseUtils.compareResultAsText(results, expected);
+    }
+
+    @Test
+    public void testBushyJoinReorderDpThreshold() throws Exception {
```

Comment:
this test is unnecessary, because it can be verified via plan test

Reply:
> this test is unnecessary, because it can be verified via plan test

Done!